public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts

```diff
@@ -0,0 +1,227 @@
+    const { uid, id } = dashboard.state;
+
+    if (this.querySub) {
+      this.querySub.unsubscribe();
+    }
+
+    if (!this.canWork(timeRange)) {
+      return;
+    }
+
+    let alerStatesExecution: Observable<AlertStateInfo[]> | undefined;
+
+    if (this.isUsingLegacyAlerting()) {
+      alerStatesExecution = from(
+        getBackendSrv().get(
+          '/api/alerts/states-for-dashboard',
+          {
+            dashboardId: id,
+          },
+          `dashboard-query-runner-alert-states-${id}`
+        )
+      ).pipe(map((alertStates) => alertStates));
+    } else {
+      alerStatesExecution = from(
+        getBackendSrv().get(
+          '/api/prometheus/grafana/api/v1/rules',
+          {
+            dashboard_uid: uid!,
+          },
+          `dashboard-query-runner-unified-alert-states-${id}`
+        )
+      ).pipe(
+        map((result: PromRulesResponse) => {
+          if (result.status === 'success') {
+            this.hasAlertRules = false;
+            const panelIdToAlertState: Record<number, AlertStateInfo> = {};
+
+            result.data.groups.forEach((group) =>
+              group.rules.forEach((rule) => {
+                if (isAlertingRule(rule) && rule.annotations && rule.annotations[Annotation.panelID]) {
+                  this.hasAlertRules = true;
+                  const panelId = Number(rule.annotations[Annotation.panelID]);
+                  const state = promAlertStateToAlertState(rule.state);
+
+                  // there can be multiple alerts per panel, so we make sure we get the most severe state:
+                  // alerting > pending > ok
+                  if (!panelIdToAlertState[panelId]) {
+                    panelIdToAlertState[panelId] = {
+                      state,
+                      id: Object.keys(panelIdToAlertState).length,
+                      panelId,
+                      dashboardId: id!,
+                    };
+                  } else if (
+                    state === AlertState.Alerting &&
+                    panelIdToAlertState[panelId].state !== AlertState.Alerting
+                  ) {
+                    panelIdToAlertState[panelId].state = AlertState.Alerting;
+                  } else if (
+                    state === AlertState.Pending &&
+                    panelIdToAlertState[panelId].state !== AlertState.Alerting &&
+                    panelIdToAlertState[panelId].state !== AlertState.Pending
+                  ) {
+                    panelIdToAlertState[panelId].state = AlertState.Pending;
+                  }
+                }
+              })
+            );
+            return Object.values(panelIdToAlertState);
+          }
+
+          throw new Error(`Unexpected alert rules response.`);
+        })
+      );
+    }
+    this.querySub = alerStatesExecution.subscribe({
+      next: (stateUpdate) => {
+        this.publishResults(
+          {
+            state: LoadingState.Done,
+            series: [toDataFrame(stateUpdate)],
```

Comment:
Returned alert states are converted to data frame and handled accordingly in the SceneQueryRunner (filtered and rediced to a single alert state).

Reply:
@dprokop I have never seen toDataFrame called with a normal object array like this, what happens?

Reply:
Object keys will be translated to fields basically and each array value to a row. Here's the implementation: https://github.com/grafana/grafana/blob/dashboard-scene-alert-states/packages/grafana-data/src/dataframe/ArrayDataFrame.ts#L28

Reply:
ah, interesting. thought that required an explicit constructor call.